File: leetcode/373.FindKPairsWithSmallestSums.cpp
// Similar to Solution. Use vector instead of pair.
//
// m is the size of nums1 and n is the size of nums2.
// Time complexity: O(min(k⋅logk,m⋅n⋅log(m⋅n)))
// Space complexity: O(min(k,m⋅n))
class Solution2 {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<vector<int>> ans;
        int n1 = nums1.size(), n2 = nums2.size();

        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> minHeap;

        minHeap.push({nums1[0] + nums2[0], 0, 0});
        set<vector<int>> visited;

        for (; k > 0 && !minHeap.empty(); -- k) {
            auto top = minHeap.top();
            int sum = top[0], i = top[1], j = top[2];
            minHeap.pop();
            ans.push_back({nums1[i], nums2[j]});
    
            if (i + 1 < n1 && !visited.contains({i + 1, j})) {
                minHeap.push({nums1[i + 1] + nums2[j], i + 1, j});
                visited.insert({i + 1, j});
            }
            if (j + 1 < n2 && !visited.contains({1, j + 1})) {
                minHeap.push({nums1[i] + nums2[j + 1], i, j + 1});
                visited.insert({i, j + 1});
            }
        }
        return ans;
    }
};


// Works. Use MinHeap.
// m is the size of nums1 and n is the size of nums2.
// Time complexity: O(min(k⋅logk,m⋅n⋅log(m⋅n)))
// Space complexity: O(min(k,m⋅n))
class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<vector<int>> ans;
        int n1 = nums1.size(), n2 = nums2.size();

        priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>,
            greater<pair<int, pair<int, int>>>> minHeap;

        minHeap.push({nums1[0] + nums2[0], {0, 0}});
        set<pair<int, int>> visited;

        for (; k > 0 && !minHeap.empty(); -- k) {
            auto top = minHeap.top();
            int sum = top.first, i = top.second.first, j = top.second.second;
            minHeap.pop();
            ans.push_back({nums1[i], nums2[j]});
    
            if (i + 1 < n1 && !visited.contains(pair<int, int>{i + 1, j})) {
                minHeap.push({{nums1[i + 1] + nums2[j]}, {i + 1, j}});
                visited.insert({i + 1, j});
            }
            if (j + 1 < n2 && !visited.contains(pair<int, int>{1, j + 1})) {
                minHeap.push({{nums1[i] + nums2[j + 1]}, {i, j + 1}});
                visited.insert({i, j + 1});
            }
        }
        return ans;
    }
};

/**
373. Find K Pairs With Smallest Sums

You are given two integer arrays nums1 and nums2 sorted in non-decreasing order and an integer k.

Define a pair (u, v) which consists of one element from the first array and one element from the second array.

Return the k pairs (u1, v1), (u2, v2), ..., (uk, vk) with the smallest sums.
 */

File: leetcode_python/373.FindKPairsWithSmallestSums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        len1, len2 = len(nums1), len(nums2)

        pairs = []
        minHeap = []
        heapq.heappush(minHeap, (nums1[0] + nums2[0], [0, 0]))
        # used = set((0, 0))  # This will be {0} instead of {{0,0}}
        used = set()
        used.add((0, 0))

        while len(minHeap) > 0 and k > 0:
            k -= 1
            sum, [x, y] = heapq.heappop(minHeap)
            pairs.append([nums1[x], nums2[y]])

            if x+1 < len1 and not (x+1, y) in used:
                used.add((x+1, y))
                heapq.heappush(minHeap, (nums1[x+1] + nums2[y], [x+1, y]))
            if y+1 < len2 and not (x, y+1) in used:
                used.add((x, y+1))
                heapq.heappush(minHeap, (nums1[x] + nums2[y+1], [x, y+1]))
            
        return pairs



/**
373. Find K Pairs with Smallest Sums
Medium

You are given two integer arrays nums1 and nums2 sorted in non-decreasing order and an integer k.

Define a pair (u, v) which consists of one element from the first array and one element from the second array.

Return the k pairs (u1, v1), (u2, v2), ..., (uk, vk) with the smallest sums.

 

Example 1:

Input: nums1 = [1,7,11], nums2 = [2,4,6], k = 3
Output: [[1,2],[1,4],[1,6]]
Explanation: The first 3 pairs are returned from the sequence: [1,2],[1,4],[1,6],[7,2],[7,4],[11,2],[7,6],[11,4],[11,6]
*/
